src/main/scala/diplomacy/LazyModule.scala

```diff
@@ -71,217 +121,386 @@ abstract class LazyModule()(implicit val p: Parameters)
+    // Preserve the previous value of the [[LazyModule.scope]], because when calling [[apply]] function,
+    // [[LazyModule.scope]] will be altered.
     val saved = LazyModule.scope
+    // [[LazyModule.scope]] stack push.
     LazyModule.scope = Some(this)
+    // Evaluate [[body]] in the current `scope`, saving the result to [[out]].
     val out = body
-    require (LazyModule.scope.isDefined, s"LazyScope ${name} tried to exit, but scope was empty!")
-    require (LazyModule.scope.get eq this, s"LazyScope ${name} exited before LazyModule ${LazyModule.scope.get.name} was closed")
+    // Check that the `scope` after evaluating `body` is the same as when we started.
+    require(LazyModule.scope.isDefined, s"LazyScope $name tried to exit, but scope was empty!")
+    require(LazyModule.scope.get eq this, s"LazyScope $name exited before LazyModule ${LazyModule.scope.get.name} was closed")
+    // [[LazyModule.scope]] stack pop.
     LazyModule.scope = saved
     out
   }
 }
 
-object LazyScope
-{
+/** Used to automatically create a level of module hierarchy (a [[SimpleLazyModule]]) within which [[LazyModule]]s can be instantiated and connected.
+  *
+  * It will instantiate a [[SimpleLazyModule]] to manage evaluation of `body`.
+  * And evaluate `body` codes in this scope.
+  */
+object LazyScope {
+  /** Create a [[LazyScope]] with name in its val name.
+    *
+    * @param body code need to be evaluated.
+    */
   def apply[T](body: => T)(implicit valName: ValName, p: Parameters): T = {
     val scope = LazyModule(new SimpleLazyModule with LazyScope)
-    scope { body }
+    scope {
+      body
+    }
   }
+
+  /** Create a [[LazyScope]] with a self defined name.
+    *
+    * @param name set valName of [[SimpleLazyModule]] created by this.
+    * @param body code need to be evaluated.
+    */
   def apply[T](name: String)(body: => T)(implicit p: Parameters): T = {
     apply(body)(ValName(name), p)
   }
 }
 
+/** One side metadata of a [[Dangle]].
+  *
+  * Describes one side of an edge going into or out of a [[BaseNode]].
+  *
+  * @param serial the global [[BaseNode.serial]] number of the [[BaseNode]] that this [[HalfEdge]] connects to.
+  * @param index  the `index` in the [[BaseNode]]'s input or output port list that this [[HalfEdge]] belongs to.
+  */
 case class HalfEdge(serial: Int, index: Int) extends Ordered[HalfEdge] {
+
   import scala.math.Ordered.orderingToOrdered
-  def compare(that: HalfEdge) = HalfEdge.unapply(this) compare HalfEdge.unapply(that)
+
+  def compare(that: HalfEdge): Int = HalfEdge.unapply(this) compare HalfEdge.unapply(that)
 }
+
+/** [[Dangle]] captures the `IO` information of a [[LazyModule]] and which two [[BaseNode]]s the [[Edges]]/[[Bundle]] connects.
+  *
+  * [[Dangle]]s are generated by [[BaseNode.instantiate]]
+  * using [[MixedNode.danglesOut]] and [[MixedNode.danglesIn]] ,
+  * [[LazyModuleImp.instantiate]] connects those that go to internal or explicit IO connections
```

Comment:
```suggestion
  * [[LazyModuleImp.instantiate]] connects those that go to internal
```

apparently the "explicit IO" is no more :(